sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1644 @@
+while keeping the old parser for processing all blocks prior to the change,
+ensures that any new behaviors introduced will only apply to transactions after
+the change takes effect.
+
+### New feature: New Trait Semantics
+
+Over the lifetime of Clarity 1, it was discovered that there are a few
+surprising behaviors of the current trait semantics.  Clarity 2 proposes the
+following trait semantics:
+
+* **No more trait type**.  The Clarity 2 type system will not distinguish between trait
+  concretizations and principals in the type system.  Consequentially, traits can be used in any way that principals
+  can.  They can be stored within lists, tuples, optionals, and responses, and
+dynamic dispatch to their methods will be permitted.
+
+* **Dynamic dispatch follows data flow**.  Because trait concretizations are
+  treated as principals, the Clarity 2 VM preserves the concretization attribute
+across data flows.  That is, when a trait concretization is bound to one or more
+variable names, dynamic dispatch to the trait concretization's functions is
+permitted on the bound variables.  This property is preserved even if the trait
+concretization is embedded within a larger structure.
+
+* **Dynamic dispatch is lost on persistence**.  Although it follows data flow, a trait
+concretization is _not_ preserved on data persistence -- the act of storing a
+principal that represents a trait concretization to a data var or a data map is
+the act of first coercing it to a bare principal.  The associated trait concretization is
+lost upon storage.  So, when loaded from storage again, the data will be a bare
+principal -- the Clarity VM will not permit code to call methods in the trait
+concretization it previosuly represented.
+
+* **Dynamic dispatch follows type compatibility**.  When type-checking a trait
+  concretization, it is sufficient that the trait concretization implements the
+functions indicated by the trait definition.  The trait concretization does
+_not_ need to implement the _same_ trait; it only needs to implement a _compatible_
+trait.
+
+#### Examples
+
+```clarity
+;; use trait in optional
+(define-public (execute (job (optional <job-trait>))) (match job j (contract-call? j fn) (default-job)))
+
+;; use of compatible traits, e.g. marketplace commissions
+(define-trait t1 ((fn (response uint bool)))) 
+(define-trait t2 ((fn (response uint bool)))) 
+```
+
+```clarity
+;; --- contract 1
+(impl-trait .contract.t1)
+;; ...
+```
+
+```clarity
+;; -- contract 2
+(impl-trait .contract.t2)
+;; ...
+```
+
+```clarity
+;; -- contract 3
+(use-trait t contract.t1)
+;; can be called with contract 1 and contract 2 because of compatible traits
+(define-public (meta-fn (ctr <t>)...)
+```
+
+### Changed: Comparators `>`, `>=`, `<=`, `<`
+
+In Clarity version 2, these binary comparators will be extended to support
+comparison of `string-ascii`, `string-utf8` and `buff`.
+
+These comparisons are done using a lexicographical comparison. 
```

Comment:
Need to clarify what *lexicographical comparison* means for `string-utf8`. Rust's [lexical_sort](https://docs.rs/lexical-sort/latest/lexical_sort/) crate says this means that _non-ASCII characters such as á or ß are treated like their closest ASCII character: á is treated as a, ß is treated as ss, etc._

Reply:
Need to clarify what lexicographical comparison means for buffers and the rationale for using it.

Reply:
Added text that describes how the comparators work on `string-utf8` and added examples.